### DISCUSSION TITLE
Itiwn performance testing

### DISCUSSION THREAD

**Discussion by bear1004** [iTwin/itwinjs-core#3702] [General] [answered]
I am a master's student at Gyeongsang National University in Korea.

I'm currently working on a performance analysis for digital twins.

While researching currently open and commercialized digital twins, I found a platform called itwin.

I want to measure the read/write performance for itwin by data size and by number of things.

Currently, an error has occurred in the process of building itwin through github, so it is not available. 
plsease, can you help me?

thank you

**Comment by pmconne** [comment]
> an error has occurred in the process of building itwin through github

What is the error? What does "building itwin through github" mean? We can help if you provide more information.

**Comment by bear1004** [comment, marked as answer]
I tried to install iTwin through Gitclone and Rush and proceed with the build by referring to "https://github.com/iTwin/itwinjs-core". 

The build was successful, but an error occurred while executing the command "rush cover".

 I have requested assistance with an error on "https://github.com/iTwin/itwinjs-core/issues/3551". 
 
But no one has helped me yet. :(

 I have one more question. Can I experiment with real-time data communication between the physical and virtual spaces I want by building iTwin via "https://github.com/iTwin/itwinjs-core"

**Reply by pmconne** [reply]
I just replied to your other question about the test errors. Ignore them for now. The tests don't have to pass for the build to succeed; you don't even have to run them.

What are you trying to achieve? I don't understand what you mean by the following - can you elaborate?
> Can I experiment with real-time data communication between the physical and virtual spaces

**Reply by bear1004** [reply]
I am currently conducting research to measure the real-time performance of digital twin. The digital twin consists of three components: physical space (real world), virtual space (simulator), and connection between physical and virtual spaces.

Therefore, we believe that real-time communication between physical and virtual spaces is important, so we want to investigate the data communication performance of traditional digital twin platforms.

**Reply by bear1004** [reply]
Physical space and virtual space are terms used in digital twin studies, so it is okay to ignore them. Ultimately, I would like to measure the data communication performance of the digital twin.

**Reply by pmconne** [reply]
The short but not very useful answer is, "yes, you can experiment".
Do you have specific questions? Have you perused the [documentation and samples](https://www.itwinjs.org)?

**Reply by bear1004** [reply]
I tried to sample before but failed. I'll try again. If an error occurs during the installation process, is it okay to ask again?

**Reply by pmconne** [reply]
Yes, just provide enough info (like the text of the error(s)) for us to assist.

**Reply by bear1004** [reply]
I appreciate it.
I'll try again. Have a nice day.

**Reply by bear1004** [reply]
@pmconne 
Hello,
I have a few questions.
1) What is the difference between the iTwin of "https://github.com/iTwin/itwinjs-core" and "https://www.itwinjs.org/"?
2) Which one should I install "https://github.com/iTwin/itwinjs-core" or "https://www.itwinjs.org/" to measure the data communication performance of my desired iTwin?
3) "https://github.com/iTwin/itwinjs-core" has been completed up to the current build. How do you communicate data with the iTwin here?

**Reply by ColinKerr** [reply]
https://www.itwinjs.org/ is the documentation site built (mostly) from the code and markdown in the itwinjs-core repository.  So when using itwinjs-core you can go to itwinjs.org for API documentation, conceptual documentation and samples.

You do not need to build the code in the itwinjs-core repository to use iTwin Platform.  You can consume the libraries built from itwinjs-core as npm packages, they are published on npmjs.com along with our other npm packages (see @itwin and @Bentley scopes)

One place to start is to create a iTwin Viewer application.  This type of application can view the 3d model and query data from the itwin.  You can create a web app using our PasS offerings using this tutorial: https://developer.bentley.com/tutorials/web-application-quick-start/. (note developer.bentley.com is the site for our paid PaaS Digital Twin services, but we offer a free tier so you can try it out) or you can create a desktop application using our itwin viewer library: https://github.com/iTwin/viewer/tree/master/packages/modules/desktop-viewer-react. 

With either type of application you can query the digital twin using 'ECSql', our query language, and use other APIs to get information about individual elements of the digital twin.  

ECSql: https://www.itwinjs.org/learning/ecsql/
Webinars on querying using ECSql:
Part 1: https://www.youtube.com/watch?v=vmQIi6RRyv8&list=PL6YCKeNfXXd9VFzTF-NSeN_l5vB7OTWR6&index=2&t=5s
Part 2: https://www.youtube.com/watch?v=aftgWdPlx1s&list=PL6YCKeNfXXd9VFzTF-NSeN_l5vB7OTWR6&index=3&t=2s
> Note you might be interested in the entire webinar series: https://www.youtube.com/playlist?list=PL6YCKeNfXXd9VFzTF-NSeN_l5vB7OTWR6

You can use the IModelConnection to make ECSql calls and other API calls to get information from the digital twin: https://www.itwinjs.org/reference/core-frontend/imodelconnection/imodelconnection/



Is this the kind of stuff you are looking for?  If not please provide more details about what you want to do

**Reply by bear1004** [reply]
What I want to do is measure real-time data communication of digital twin platform according to data. 3D modeling is not required in my experiment.

I want to measure the speed at which I store custom data on a digital twin platform and extract the stored data.

 Are you saying I can measure using "ECSql" in "ITwin Viewer"?

**Reply by ColinKerr** [reply]
The iTwin Viewer can only query/view data in an [iModel (the central Db for an iTwin)](https://www.itwinjs.org/learning/imodels/).  To write data to an iModel you have to use our backend APIs: https://www.itwinjs.org/learning/backend/. 

Here is a very simple test app that inserts data into a local iModel: https://github.com/iTwin/itwinjs-core/blob/master/test-apps/imjs-importer/src/ImportIMJS.ts

One of the big strengths of the iTwin Platform is that an iModel can be [synchronized](https://www.itwinjs.org/learning/backend/imodeldbsync/) between multiple computers using our [iModels API](https://developer.bentley.com/apis/imodels/).  We offer [synchronization services that automatically synchronizes data from Engineering Application files to an iModel](https://developer.bentley.com/apis/synchronization/).  You can try this out by following the instructions in our [Getting Started](https://www.itwinjs.org/getting-started/#2-get-a-sample-imodel) page.

**Reply by bear1004** [reply]
Thank you so much for giving me so much information!!!
I'll look at the documents and try.

**Reply by bear1004** [reply]
@pmconne @ColinKerr 
I ran into an error while doing the tutorial.
"https://www.itwinjs.org/learning/tutorials/develop-web-viewer/"

I created the application and IModel in the order they appear in the documentation. After that, the ".env" file was modified.

After executing the "npm start" command, http://localhost:3000/ was accessed. But I got the error "The user is unauthorized. Please provide valid authentication credentials".

please, can you help?
